src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,411 @@
+        }
+    }
+
+    /**
+     * Uploads a file from disk.
+     *
+     * The destination bucket should reside in the same region as the one configured for the client.
+     *
+     * Pipes the file (read) stream into the request (write) stream.
+     * Assigns the target content type based on the mime type of the file.
+     * If content type cannot be determined, defaults to {@link DEFAULT_CONTENT_TYPE}.
+     *
+     * @throws S3Error if there is an error calling S3 or piping between streams.
+     */
+    public async uploadFile(request: UploadFileRequest): Promise<void> {
+        const s3 = await this.createS3()
+
+        // https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/s3-example-creating-buckets.html#s3-example-creating-buckets-upload-file
+        const readStream = this.fileStreams.createReadStream(request.fileLocation)
+        const contentType = mime.lookup(basename(request.fileLocation.fsPath)) || DEFAULT_CONTENT_TYPE
+
+        let managedUploaded = s3.upload({
+            Bucket: request.bucketName,
+            Key: request.key,
+            Body: readStream,
+            ContentType: contentType,
+        })
+
+        const progressListener = request.progressListener
+        if (progressListener) {
+            managedUploaded.on('httpUploadProgress', progress => {
+                progressListener(progress.loaded)
+            })
+        }
+
+        try {
+            await Promise.all([promisifyReadStream(readStream), managedUploaded.promise()])
+        } catch (e) {
+            throw new S3Error(`Failed to upload ${request.key} to bucket ${request.bucketName}: ${e}`)
+        }
+    }
+
+    /**
+     * Lists buckets in the region of the client.
+     *
+     * Note that S3 returns all buckets in all regions,
+     * so this incurs the cost of additional S3#HeadBucket requests for each bucket
+     * to filter out buckets residing outside of the client's region.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    public async listBuckets(): Promise<ListBucketsResponse> {
+        const s3 = await this.createS3()
+
+        let s3Buckets: S3.Bucket[]
+        try {
+            const output = await s3.listBuckets().promise()
+            s3Buckets = output.Buckets ?? []
+        } catch (e) {
+            throw new S3Error(`Failed to list buckets: ${e}`)
+        }
+
+        // S3#ListBuckets returns buckets across all regions
+        const allBucketPromises: Promise<Bucket>[] = s3Buckets.map(
+            async s3Bucket =>
+                new DefaultBucket({
+                    partitionId: this.partitionId,
+                    region: await this.lookupRegion(s3Bucket.Name!, s3),
+                    name: s3Bucket.Name!,
+                })
+        )
+
+        const allBuckets = await Promise.all(allBucketPromises)
+        const bucketsInRegion = _(allBuckets)
```

Comment:
Would there ever be a scenario where we want to return allBuckets? e.g. if we're refreshing every region, we'd have to make a listBuckets call and fully filter the list for each region we have open.

Reply:
Yeah, it might be good to support listing all buckets and listing a region's buckets

Reply:
Adding a better example: our deploy Lambda workflow (refreshing the explorer doesn't make as much sense since we collapse everything on refresh).

Reply:
I agree that it's redundant and not entirely optimal. However, I think it's actually more confusing to have a region-specific client that's able to return things from other regions.

I don't think the filtering is a big deal because there's a strict limit to number of buckets in the account and this seems fast. How about we revisit it if it actually turns down to be problematic? I prefer the consistency of a region-specific client over the (possible) performance gain.